### PR TITLE
asusctl, supergfxctl: init, add nixos modules

### LINKED ISSUE #131811: Asusctl
**Project description**
It is a command line utility which adds features for asus laptops. For example graphics card switching, anime matrix control (for g14). fan control, keyboard backlight control, cpu boost control, profiles, fan curves and other stuff. It basically unlocks a lot of functionality for laptops and can be used for non asus laptops as well contrary to the name.

**Metadata**

* homepage URL:https://asus-linux.org/asusctl/
* source URL:https://gitlab.com/asus-linux/asusctl
* license: Mozilla Public License 2.0
* platforms: Linux

### PR DESCRIPTION
**Pull request by aacebedo**
###### Description of changes

Adding the asusctl package.
It includes several daemons and a tool to configure several aspects of the Asus ROG line of computers.
See https://asus-linux.org/

First time I package a service so I may have done some mistakes.

Supersedes https://github.com/NixOS/nixpkgs/pull/177940 https://github.com/NixOS/nixpkgs/pull/147786 https://github.com/NixOS/nixpkgs/pull/146751
Closes https://github.com/NixOS/nixpkgs/issues/131811

###### Things done

- Built on platform(s)
  - [ x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/module-list.nix
Comment (SuperSandro2000) [inline nixos/modules/module-list.nix:43]: This is not sorted properly

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:42]:
```suggestion
          file = "${cfg.package}/share/asusd/anime/custom/sonic-run.gif";
```
Can we drop this from the package?

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:14]: From where did you get this? How did you generate it? Should we move it into an extra file?

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:143]:
```suggestion
        default = false;
        description = lib.mdDoc ''
          Activate the asusd-user service.
        '';
```
normally we put default before description

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:149]: This should mention that this is a configuration file and where to discover what you can write into that. Is there a doc we can link?

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:175]:
```suggestion
        default = "";
```

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:184]: This default is way way to long and also If I would want to append one thing I would need to copy this all.

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:391]: Would you want to turn this of normally?

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:406]:
```suggestion
```

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:412]: Can we use some condition instead?

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:418]: NixOS does not support SeLinux.

```suggestion
```

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:439]:
```suggestion
      wantedBy = [ "multi-user.target" ];
```

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:470]: Can't we leave the default mode? The location is not writtable anyway

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:476]: Can't we add the package to udev.packages?

File: pkgs/applications/system/asusctl/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/system/asusctl/default.nix:19]:
```suggestion
  version = "4.5.0";

```

File: pkgs/applications/system/asusctl/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/system/asusctl/default.nix:39]: src is normally put between version and nativeBuildInputs

File: pkgs/applications/system/asusctl/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/system/asusctl/default.nix:54]: What crimes are they doing with chattr?

File: pkgs/applications/system/asusctl/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/system/asusctl/default.nix:53]:
```suggestion
    for file in daemon/src/config.rs daemon-user/src/daemon.rs daemon-user/src/ctrl_anime.rs daemon-user/src/user_config.rs rog-control-center/src/main.rs; do
      sed -i -e s,/usr/share,$out/share, $file
    done
```

or can sed do multiple files natively? Not sure if this is better or not.

File: pkgs/applications/system/asusctl/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/system/asusctl/default.nix:88]: Is there no install script? Can we shrink this with wildcards or functions?

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:38066]: Please sort this.

Comment (aacebedo) [reply pkgs/top-level/all-packages.nix:38066]: Where shall I put it exactly ? I don't see the logic in the all-packages.nix as everything is not sorted.

File: pkgs/applications/system/asusctl/default.nix
Comment (aacebedo) [reply pkgs/applications/system/asusctl/default.nix]: Factorized using *.
The makefile is not usable because they install other things like the service and rules. I think it is better to do this in the nix file instead of patching the makefile.

File: pkgs/applications/system/asusctl/default.nix
Comment (aacebedo) [reply pkgs/applications/system/asusctl/default.nix]: Done

File: pkgs/applications/system/asusctl/default.nix
Comment (aacebedo) [reply pkgs/applications/system/asusctl/default.nix:61]: They activate/deactivate the sound made by the machine when the BIOS is initialized.

File: pkgs/applications/system/asusctl/default.nix
Comment (aacebedo) [reply pkgs/applications/system/asusctl/default.nix]: Done

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix:35]: Added a link to the website describing this.

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix]: Done

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix]: Found it from the first start of the binary, I removed it as it is better to rely on the default of the binary.

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix]: Done I removed the default.

File: pkgs/applications/system/asusctl/default.nix
Comment (aacebedo) [reply pkgs/applications/system/asusctl/default.nix:19]: done

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix:167]: Unfortunately no. The daemon writes those files on request by the rog-control-center binary provided in the package modifies those files when it is started.

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix]: Done

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix]: Ok I removed it.

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix]: copied it from the repo but agreed that is is not good so I removed it.

File: nixos/modules/services/hardware/asusctl.nix
Comment (aacebedo) [reply nixos/modules/services/hardware/asusctl.nix]: removed

**Comment by aacebedo** [comment]
Question about the merge process. When suggestions for modifications are accepted, it creates a commit.

However I am trying to keep a single commit for the new package as it is a good practice used on the project. 

Shall I squash all commits into a single one and keep the co-authoring comments into it ? I was amending the commit but I cannot keep the co-authoring mentions on it and I think it would be better to keep it.

**Comment by aacebedo** [comment]
@SuperSandro2000 can you give a second look to the MR please ? Thanks !

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:38066]: Try to find ``### APPLICATIONS`` or sort it near other applications but appending it provokes merge conflicts.

**Comment by SuperSandro2000** [comment]
> Shall I squash all commits into a single one and keep the co-authoring comments into it ? I was amending the commit but I cannot keep the co-authoring mentions on it and I think it would be better to keep it.

You can also do the changes locally and ammend them. Keeping the Co-Author for review suggestions is not important.

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:123]: Why are we not using multi-user.target here?

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/asusctl.nix:11]:
```suggestion
  animeConfigFile = pkgs.writeText "anime.conf" (lib.optionalString (cfg.animeConfig != { }) builtins.toJSON cfg.animeConfig);
  asusdConfigFile = pkgs.writeText "asusd.conf" (lib.optionalString (cfg.asusdConfig != { }) builtins.toJSON cfg.asusdConfig);
  auraConfigFile = pkgs.writeText "aura.conf" (lib.optionalString (cfg.auraConfig != { }) builtins.toJSON cfg.auraConfig);
  profileConfigFile = pkgs.writeText "profile.conf" (lib.optionalString (cfg.profileConfig != "") builtins.toJSON cfg.profileConfig);
```

File: nixos/modules/services/hardware/asusctl.nix
Comment (SuperSandro2000) [reply nixos/modules/services/hardware/asusctl.nix:167]: I am not sure if you can write to those files even with 644. Also nix overwrites them when switching/restarting.

**Comment by K900** [comment]
Previous attempts, for posterity: https://github.com/NixOS/nixpkgs/pull/177940 https://github.com/NixOS/nixpkgs/pull/147786 https://github.com/NixOS/nixpkgs/pull/146751

File: nixos/modules/services/hardware/asusctl.nix
Comment (K900) [reply nixos/modules/services/hardware/asusctl.nix:167]: You definitely can't.

File: pkgs/applications/system/asusctl/default.nix
Comment (K900) [inline pkgs/applications/system/asusctl/default.nix:89]: Does this not get handled automatically? Also, `--add-rpath`.

Comment (aacebedo) [reply pkgs/applications/system/asusctl/default.nix:89]: Unfortunately not, if the patch is not made the control center does not work.

**Comment by aacebedo** [comment]
> Previous attempts, for posterity: #177940 #147786 #146751

Oh interesting.
Shall I close this MR and retake the one of the other ones ? Why didn't those attempts end up in the master branch ?

File: nixos/modules/services/hardware/asusctl.nix
Comment (K900) [inline nixos/modules/services/hardware/asusctl.nix:8]: This breaks eval now. Should be `(lib.optionalString (condition) (builtins.toJSON cfg.animeConfig))`

**Comment by K900** [comment]
> > Previous attempts, for posterity: #177940 #147786 #146751
> 
> Oh interesting. Shall I close this MR and retake the one of the other ones ? Why didn't those attempts end up in the master branch ?

You can't, there's no way to really take over a PR unless you have the commit bit.

**Comment by aacebedo** [comment]
> > > Previous attempts, for posterity: #177940 #147786 #146751
> > 
> > 
> > Oh interesting. Shall I close this MR and retake the one of the other ones ? Why didn't those attempts end up in the master branch ?
> 
> You can't, there's no way to really take over a PR unless you have the commit bit.

Ok at least I will take a look to the previous PRs and see if some stuff are written in a better way. 
I really need this service to use my machine efficiently so I hope this PR will succeed this time.

**Comment by K900** [comment]
Looks like we also need to build/install/enable supergfxctl for the GUI to work at least on my laptop (GA402RJ).

**Comment by aacebedo** [comment]
> Looks like we also need to build/install/enable supergfxctl for the GUI to work at least on my laptop (GA402RJ).

yes. I did it on my laptop already and found a bug in the soft. I am going to push the PR tonight

**Review by K900 — CHANGES_REQUESTED** [review]
Preventative r- so this doesn't get merged until we have supergfxd running.

**Comment by K900** [comment]
Seems to be working now.

**Comment by aacebedo** [comment]
Modifications looks good. Didnt know some of thi techniques you use for the replacement in files. 
Didnt know that we shall use the original service file instead of recreating it through nix commands. 

Shall we split the MR in two (one for supergfxctl)? 

Can you also tell me if supergfxctl always reports the gpu as suspended? On my machine it is unable to differentiate the gpu with the HDMI/DP controller and report an incorrect status.

**Comment by K900** [comment]
Upstream broke the build again, waiting for response on https://gitlab.com/asus-linux/asusctl/-/issues/283

**Comment by aacebedo** [comment]
tested on my side yesterday and confirmed it was working. Are you ok if I squash commits into 2, one for asusctl and one for supergfxctl.

**Comment by K900** [comment]
I'd rather see if upstream responds first.

**Comment by K900** [comment]
Cleaned up the history and updated to new upstream point release so it builds without cargo.lock hacks.

**Review by K900 — APPROVED** [review]
This should be good to go now.

**Comment by bobvanderlinden** [comment]
Good to see someone else also stood up to package this :+1: Thanks for creating the PR and others for reviewing! I previously created a PR but because I do not own a compatible asus laptop/card it was hard to do maintenance.

If you'd like to follow-up on this, it might be nice to integrate this into nixos-hardware to have a configuration specific to your model of laptop.

**Comment by K900** [comment]
Can't comment on other devices, but my GA402R works perfectly with no special configuration required, so a nixos-hardware module doesn't seem necessary.

**Comment by aacebedo** [comment]
I have the same device but I use a custom profile for nixos-hardware. I'll
commit it then.

**Comment by K900** [comment]
Please ping me on the PR if you do.

**Comment by icyphox** [comment]
Very minor nit: wouldn't it make more sense naming the services asusd and supergfxd, as opposed to *ctl?

**Comment by aacebedo** [comment]
That's a good point.
When I created the asusctl service I used this name because the project is
called this on gitlab. However I agree that we can change at least the name
of the service because in the repo the service  file is named asusd too.

**Comment by aacebedo** [comment]
Renamed services as asusd and supergfxd. Also separated rog-control-center accordingly

**Comment by K900** [comment]
One more cleanup pass, hopefully the last one: limited `meta.platforms` to `x86_64-linux` since this is not useful outside of those targets anyway + a few formatting nits. Will probably merge this once ofborg is done, we can fix up the nits later if necessary.

File: nixos/modules/services/hardware/supergfxd.nix
Comment (dywedir) [inline nixos/modules/services/hardware/supergfxd.nix:26]: Should be
```
environment.etc."supergfxd.conf".source = #...
```

Comment (mtoohey31) [reply nixos/modules/services/hardware/supergfxd.nix:26]: Also, I may be misreading [the source code](https://gitlab.com/asus-linux/supergfxctl/-/blob/5.0.1/src/config.rs#L100), but I believe the format should be JSON, not INI.

**Comment by andOrlando** [comment]
Is supergfxd starting for you guys? I'm getting `thread 'main' panicked at 'The directory /etc/supergfxd.conf is missing', src/config.rs:94:33`. /etc/supergfxd.conf does exist, though it's linked /nix/store which is a read-only system and it tries to open with read, write and create permissions which is most likely the issue.

Let me know if there's a better place to bring this up

**Comment by mtoohey31** [comment]
I didn't run into that exact issue myself, but in order to create the config file with the correct format (instead of INI) I had to add:

```nix
  environment.etc."supergfxd.conf" = {
    mode = "0644";
    source = (pkgs.formats.json { }).generate "supergfxd.conf" {
      mode = "Integrated";
      vfio_enable = true;
      vfio_save = false;
      always_reboot = false;
      no_logind = true;
      logout_timeout_s = 180;
      hotplug_type = "Asus";
    };
  };
```

I think the `mode` attribute in particular might what you need to add @andOrlando, so that the file is writeable. You'll probably want to change the contents of the config file to match your desired settings, just make sure that you provide values for all the fields, or else supergfxd will fail to deserialize it and will re-create it with the default values.

Also, I'll leave this here too in case anyone finds it useful, I also had to include the following since I'm using `hotplug_type = "Asus"`:

```nix
  systemd.services.supergfxd.path = [ pkgs.kmod pkgs.pciutils ];
```

**Pull request merged**
